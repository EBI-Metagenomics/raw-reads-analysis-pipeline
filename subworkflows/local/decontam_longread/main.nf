include { MINIMAP2_ALIGN } from '../../../modules/nf-core/minimap2/align/main'

workflow DECONTAM_LONGREAD {
    take:
    input_reads // [ val(meta), path(reads) ]
    reference_genome // [ val(meta2), path(reference_genome) ]

    main:
    def ch_versions = Channel.empty()

    MINIMAP2_ALIGN(
        input_reads,
        reference_genome,
        "human",
        "fastq",
        true,
        "bai",
        false,
        true,
    )
    ch_versions = ch_versions.mix(MINIMAP2_ALIGN_HUMAN.out.versions)

    decontaminated_reads = MINIMAP2_ALIGN_HUMAN.out.filtered_output

    emit:
    qc_reads = decontaminated_reads
    versions = ch_versions
}
