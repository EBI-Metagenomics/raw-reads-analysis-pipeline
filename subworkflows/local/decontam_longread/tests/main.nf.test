nextflow_workflow {

    name "Test Subworkflow DECONTAM_LONGREAD"
    script "../main.nf"
    workflow "DECONTAM_LONGREAD"

    tag "subworkflows"
    tag "subworkflows_"
    tag "subworkflows/decontam_longread"
    tag "minimap2"
    tag "minimap2/align"

    test("mini ONT sample output should match snapshot") {

        when {
	    params {
	        remove_phix = true
	    }
            workflow {
                """
                input[0] = Channel.of([
		    ['id': 'test', 'single_end': true, 'platform': 'ont'],
	            file("${projectDir}/tests/fixtures/decontam_shortread/ERR10889147_mini_1.fq.gz", checkIfExists: true),
		])
		input[1] = Channel.of([
		    ['id': 'host_genome_minimap2', 'base_dir': '.', 'files': ['index': 'hg38_mini.mmi']],
		    file("${projectDir}/tests/fixtures/dbs/hg38_mini_minimap2", checkIfExists: true)
		])
                """
            }
        }

        then {
	    def reads = path(workflow.out.decontaminated_reads.get(0)[1])
	        .linesGzip
		.collect{
	            if(it.charAt(0)=="@"){
	                return it
	 	    }
	        }
		.findAll{ it != null }
	        .sort()

            assert workflow.success
            assert snapshot(
                reads, 
		workflow.out.versions
	    ).match()
        }

    }

    test("mini PB sample output should match snapshot") {

        when {
	    params {
	        remove_phix = true
	    }
            workflow {
                """
                input[0] = Channel.of([
		    ['id': 'test', 'single_end': true, 'platform': 'pb'],
	            file("${projectDir}/tests/fixtures/decontam_shortread/ERR10889147_mini_2.fq.gz", checkIfExists: true),
		])
		input[1] = Channel.of([
		    ['id': 'host_genome_minimap2', 'base_dir': '.', 'files': ['index': 'hg38_mini.mmi']],
		    file("${projectDir}/tests/fixtures/dbs/hg38_mini_minimap2", checkIfExists: true)
		])
                """
            }
        }

        then {
	    def reads = path(workflow.out.decontaminated_reads.get(0)[1])
	        .linesGzip
		.collect{
	            if(it.charAt(0)=="@"){
	                return it
	 	    }
	        }
		.findAll{ it != null }
	        .sort()

            assert workflow.success
            assert snapshot(
                reads, 
		workflow.out.versions
	    ).match()
        }

    }

    test("mini ONT sample with added hg38 synthetic reads sample output should match snapshot") {

        when {
	    params {
	        remove_phix = true
	    }
            workflow {
                """
                input[0] = Channel.of([
		    ['id': 'test', 'single_end': true, 'platform': 'ont'],
		    file("${projectDir}/tests/fixtures/decontam_shortread/ERR10889147_mini_plus_hg38_1.fq.gz", checkIfExists: true),
		])
		input[1] = Channel.of([
		    ['id': 'host_genome_minimap2', 'base_dir': '.', 'files': ['index': 'hg38_mini.mmi']],
		    file("${projectDir}/tests/fixtures/dbs/hg38_mini_minimap2", checkIfExists: true)
		])
                """
            }
        }

        then {
	    def reads = path(workflow.out.decontaminated_reads.get(0)[1])
	        .linesGzip
		.collect{
	            if(it.charAt(0)=="@"){
	                return it
	 	    }
	        }
		.findAll{ it != null }
	        .sort()

            assert workflow.success
            assert snapshot(
                reads, 
		workflow.out.versions
	    ).match()
        }

    }

    test("hg38 synthetic ONT reads sample output should match snapshot") {

        when {
	    params {
	        remove_phix = true
	    }
            workflow {
                """
                input[0] = Channel.of([
		    ['id': 'test', 'single_end': true, 'platform': 'ont'],
		    file("${projectDir}/tests/fixtures/decontam_shortread/hg38_reads_1.fq.gz", checkIfExists: true),
		])
		input[1] = Channel.of([
		    ['id': 'host_genome_minimap2', 'base_dir': '.', 'files': ['index': 'hg38_mini.mmi']],
		    file("${projectDir}/tests/fixtures/dbs/hg38_mini_minimap2", checkIfExists: true)
		])
                """
            }
        }

        then {
            assert workflow.success
	    assert workflow.out.decontaminated_reads.size()==0
            assert snapshot(workflow.out.versions).match()
        }

    }

}
