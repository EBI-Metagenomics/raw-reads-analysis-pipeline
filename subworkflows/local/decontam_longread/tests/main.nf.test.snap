{
    "mini ONT sample with added hg38 synthetic reads sample output should match snapshot": {
        "content": [
            [
                "@ERR10889147.130806",
                "@ERR10889147.150910",
                "@ERR10889147.280640",
                "@ERR10889147.350555",
                "@ERR10889147.377961",
                "@ERR10889147.421340",
                "@ERR10889147.453562",
                "@ERR10889147.483471",
                "@ERR10889147.547797",
                "@ERR10889147.55288",
                "@ERR10889147.662821",
                "@ERR10889147.77699",
                "@ERR10889147.80480"
            ],
            [
                "versions.yml:md5,1acf0fee10591fb9de26449d2bcc32a6",
                "versions.yml:md5,9cbaa7bfcef359f47b9565b7d0eb999e",
                "versions.yml:md5,9cbaa7bfcef359f47b9565b7d0eb999e",
                "versions.yml:md5,9cbaa7bfcef359f47b9565b7d0eb999e",
                "versions.yml:md5,9cbaa7bfcef359f47b9565b7d0eb999e"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-05-28T12:56:43.140639024"
    },
    "hg38 synthetic ONT reads sample output should match snapshot": {
        "content": [
            [
                "versions.yml:md5,1acf0fee10591fb9de26449d2bcc32a6",
                "versions.yml:md5,9cbaa7bfcef359f47b9565b7d0eb999e",
                "versions.yml:md5,9cbaa7bfcef359f47b9565b7d0eb999e",
                "versions.yml:md5,9cbaa7bfcef359f47b9565b7d0eb999e",
                "versions.yml:md5,9cbaa7bfcef359f47b9565b7d0eb999e"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-05-28T12:58:01.737160022"
    },
    "mini PB sample output should match snapshot": {
        "content": [
            [
                "@B@BB>GEGGGEE1>FDGC@1=EBFGDC@:B<E@>F>@@CFGCGE:B<C@:00=FGG0C=1<EFGE::<@E>FGGEE@:FG0C1FGD9EG/CD00F@D8D0FCFDDGG@088=FDGGCG/CE@",
                "@ERR10889147.101689",
                "@ERR10889147.130806",
                "@ERR10889147.150910",
                "@ERR10889147.239847",
                "@ERR10889147.280640",
                "@ERR10889147.313339",
                "@ERR10889147.350555",
                "@ERR10889147.374007",
                "@ERR10889147.377961",
                "@ERR10889147.398989",
                "@ERR10889147.421340",
                "@ERR10889147.421831",
                "@ERR10889147.453562",
                "@ERR10889147.483471",
                "@ERR10889147.541000",
                "@ERR10889147.547797",
                "@ERR10889147.55288",
                "@ERR10889147.56958",
                "@ERR10889147.63497",
                "@ERR10889147.646304",
                "@ERR10889147.65722",
                "@ERR10889147.662821",
                "@ERR10889147.76311",
                "@ERR10889147.77699",
                "@ERR10889147.80480"
            ],
            [
                "versions.yml:md5,1acf0fee10591fb9de26449d2bcc32a6",
                "versions.yml:md5,9cbaa7bfcef359f47b9565b7d0eb999e",
                "versions.yml:md5,9cbaa7bfcef359f47b9565b7d0eb999e",
                "versions.yml:md5,9cbaa7bfcef359f47b9565b7d0eb999e"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-05-28T12:54:37.003929091"
    },
    "mini ONT sample output should match snapshot": {
        "content": [
            [
                "@B@BCEG>FCG>GGGG@E99CEGBGB/BEC@G@FF@@G11C1EGGA9CFGGGGCF@=FCFGGGGGEGGGGG=DGB@/FGCGBC9EEEGG/9C/EGGGGGG/8E<.E@D/6.CGGC=D/C@CE6D6",
                "@BBBBFFGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGFGGGGGGGGEGGGGGGGGGCFGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG",
                "@ERR10889147.101689",
                "@ERR10889147.130806",
                "@ERR10889147.150910",
                "@ERR10889147.239847",
                "@ERR10889147.280640",
                "@ERR10889147.313339",
                "@ERR10889147.350555",
                "@ERR10889147.374007",
                "@ERR10889147.377961",
                "@ERR10889147.398989",
                "@ERR10889147.421340",
                "@ERR10889147.421831",
                "@ERR10889147.453562",
                "@ERR10889147.483471",
                "@ERR10889147.541000",
                "@ERR10889147.547797",
                "@ERR10889147.55288",
                "@ERR10889147.56958",
                "@ERR10889147.63497",
                "@ERR10889147.646304",
                "@ERR10889147.65722",
                "@ERR10889147.662821",
                "@ERR10889147.76311",
                "@ERR10889147.77699",
                "@ERR10889147.80480"
            ],
            [
                "versions.yml:md5,1acf0fee10591fb9de26449d2bcc32a6",
                "versions.yml:md5,9cbaa7bfcef359f47b9565b7d0eb999e",
                "versions.yml:md5,9cbaa7bfcef359f47b9565b7d0eb999e",
                "versions.yml:md5,9cbaa7bfcef359f47b9565b7d0eb999e",
                "versions.yml:md5,9cbaa7bfcef359f47b9565b7d0eb999e"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-05-28T12:51:53.117587708"
    }
}